### ISSUE TITLE
Docker daemon.json "ip" parameter has no effect

### ISSUE THREAD

**Issue by archmangler** [docker/for-linux#370]
* [x] This is a bug report
* [ ] This is a feature request
* [x] I searched existing issues before opening this one

### Expected behavior

When using the "ip" parameter to specify a bind address in /etc/docker/daemon.json, the containers should appear as listening on the specified interface address.

### Actual behavior

Even with specifying "ip": <ip-address>, after restarting the docker daemon, containers show up as listening on "0.0.0.0".

### Steps to reproduce the behavior

1. Install docker 18.03.1-ce, build 9ee9f40 or higher on Ubuntu 16.04 with kerne 4.13.0-37-generic or higher.
2. Configure the static IP address in /etc/docker/daemon.json: "ip": "w.x.y.z",
3. Restart the docker daemon
4 Optional: stop/start individual containers

**Output of `docker version`:**

```
18.03.1-ce, build 9ee9f40
```

**Output of `docker info`:**

```
Containers: 4
 Running: 4
 Paused: 0
 Stopped: 0
Images: 7
Server Version: 18.03.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: syslog
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.13.0-37-generic
Operating System: Ubuntu 16.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.791GiB
Name: xx.yy.zz
ID: VSUY:ZYVM:I4G4:XDML:CC24:KRY6:RL4S:7NJS:A76W:JP27:UOR7:ESYH
Docker Root Dir: /mnt/data/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.)**

VMware VM

**Comment by ellisonpatterson** [comment]
Same for me as well.

**Comment by btgoodwin** [comment]
Any work-arounds for this?

**Comment by hyshka** [comment]
This is affecting me as well.

**Output of `docker version`:**
```
Client:
 Version:           19.03.5-ce
 API version:       1.40
 Go version:        go1.13.4
 Git commit:        633a0ea838
 Built:             Fri Nov 15 03:19:09 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          19.03.5-ce
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.4
  Git commit:       633a0ea838
  Built:            Fri Nov 15 03:17:51 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.3.1.m
  GitCommit:        d50db0a42053864a270f648048f9a8b4f24eced3.m
 runc:
  Version:          1.0.0-rc9
  GitCommit:        d736ef14f0288d6993a1845745d6756cfc9ddd5a
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker info`:**

```
Client:
 Debug Mode: false

Server:
 Containers: 5
  Running: 5
  Paused: 0
  Stopped: 0
 Images: 52
 Server Version: 19.03.5-ce
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: d50db0a42053864a270f648048f9a8b4f24eced3.m
 runc version: d736ef14f0288d6993a1845745d6756cfc9ddd5a
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
  userns
 Kernel Version: 5.3.11-x86_64-linode131
 Operating System: Arch Linux
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 3.828GiB
 Name: linode
 ID: 5C6O:KE4U:6MCR:HMCF:VBAT:D6S3:4CXI:OVEO:M4BG:WJV6:IE3J:EXDK
 Docker Root Dir: /var/lib/docker/1000.1000
 Debug Mode: true
  File Descriptors: 80
  Goroutines: 95
  System Time: 2019-12-03T09:30:06.909166365-07:00
  EventsListeners: 0
 Username: hyshka
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.)**

ArchLinux on a VPS

**Comment by arjenlentz** [comment]
Issue (still) exists on Ubuntu 18.04.4 (Docker version 19.03.6, build 369ce74a3c) as well.
Surely this can be fixed.

**Comment by cpuguy83** [comment]
I reproduced this.
This seems to be due to the issue described here: https://github.com/moby/moby/issues/40789

TLDR;
Your container is not responding to signals as it should (because `/bin/sh` doesn't respond to signals when it's run to PID1 (similar for many programs).
Docker, on startup, is supposed to kill all running containers but it is not doing this properly for the case where the container is not responding to signals appropriately and as such stays running.

Please try running the container with `--init` (you can also set `"init": true` in daemon.json to make it the default) and see if that clears it up for you.
What `--init` does is add a proper PID1 to your container that handles signals... in turn this also makes your program run as something other than PID1 and subsequently it will handle signals as expected.

This happens to be fixed by a PR we merged today: https://github.com/moby/moby/pull/40791
I haven't decided if this should be backported or not, I'm leaning on not b/c it's potentially surprising to some people (even if it's making Docker do the right thing as it once used to) and relatively simple to work-around.

**Comment by cpuguy83** [comment]
I am going to close this for housekeeping purposes because this is the only way I've been able to reproduce the issue and the fix for it is merged.

Feel free to discuss or let me know if this doesn't work for you and we can re-open.
Thank you!.

**Comment by SvenDowideit** [comment]
@cpuguy83 I'm confused - you talk about signals, but this issue is to do with port mapping to a specified IP other than 0.0.0.0 (the default)

**Comment by cpuguy83** [comment]
Yes. If the container doesn't shutdown then it will still be mapped to the old address.

**Comment by arjenlentz** [comment]
@cpuguy83 thanks for looking at this Brian. Indeed your solution appears to work, but so far I've only managed it on a special test VM. None of the regular dev environments appear to play along.

`/etc/docker/daemon.json
{
  "init": true,
  "ip": "127.0.0.1"
}`

docker run -d -p 80:80 nginx
docker ps

When it works correctly, it should show 127.0.0.1:80->80/tcp instead of 0.0.0.0:80.

**Comment by arjenlentz** [comment]
@cpuguy83 with your guru expertise: from an existing system that exhibits the problem, do you have a recommended set of steps to resolve? Killing all containers manually then restarting the daemon (with new config) does not appear to do the trick.

**Comment by cpuguy83** [comment]
Do the containers have a restart policy? Are they being restarted under the same daemon config?

**Comment by arjenlentz** [comment]
The compose files tend to have "restart: always".
What would control whether the containers are being restarted under the same (old?) daemon config rather than the new one?

**Comment by cpuguy83** [comment]
It depends on how you stopped the containers, I suppose.

**Comment by arjenlentz** [comment]
indeed - so do you have a recommendation on how to stop them in this case, to ensure they pick up the new daemon config?

**Comment by cpuguy83** [comment]
`docker stop` ought to work.

**Comment by arjenlentz** [comment]
Tried "docker stop $(docker ps -q)", verified that all containers are gone, then "systemctl restart docker", but after that still containers get started with 0.0.0.0 binding.
Hence my question, it seems to be really hard to get the change through on an existing dev box.

**Comment by cpuguy83** [comment]
Can you paste what's in your daemon.json?

**Comment by arjenlentz** [comment]
Just
```
{
  "init": true,
  "ip": "127.0.0.1"
}
```

New "docker run" tests seem to come through with the new config, but docker-compose controlled ones do not? They have ports definitions such as - 80:80 so I would expect it to use the default binding then, which should be 127.0.0.1 after the daemon config change. Right?

**Comment by arjenlentz** [comment]
k so the remaining issue appears to be docker-compose rather than docker/dockerd now, I think it's covered in https://github.com/docker/compose/issues/2999 which is still open and hangin' around as a kind-of-bug.

**Comment by cpuguy83** [comment]
Ok , I see. `docker-compose` creates, by default, it's own network rather than using the default bridge, meanwhile the configurations on the daemon tend to only apply to containers in the default bridge network... however this seems not quite right since port publishing has nothing to do with networks.

/cc @arkodg

**Comment by cpuguy83** [comment]
Re-opening because there does indeed seem to be an issue with how custom networks and port publishing interact.

**Comment by arkodg** [comment]
Youre right @cpuguy83 there is a difference between the IP:Port bindings between default bridge `bridge` and custom bridges

```
dockerd -D --ip 127.0.0.1
docker network create -d bridge testnet
docker run -d -p 80:80 --name c-bridge nginx
docker run -d -p 8080:8080 --name c-testnet --net testnet nginx
docker ps     
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                            NAMES
01271898eb02        nginx               "nginx -g 'daemon ..."   4 seconds ago       Up 3 seconds        80/tcp, 0.0.0.0:8080->8080/tcp   c-testnet
c635379339d3        nginx               "nginx -g 'daemon ..."   2 minutes ago       Up 2 minutes        127.0.0.1:80->80/tcp             c-bridge
```

This is probably due the difference in `netOptions` passed by the daemon to `libnetwork`

```
docker network inspect bridge          
[
    {
        "Name": "bridge",
        "Id": "ad71140bffd23bea56b8edf66fab0902b801263cec2ea520656cc605339d23d2",
        "Created": "2020-04-17T00:51:49.0086785Z",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "172.18.0.0/16"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {},
        "Options": {
            "com.docker.network.bridge.default_bridge": "true",
            "com.docker.network.bridge.enable_icc": "true",
            "com.docker.network.bridge.enable_ip_masquerade": "true",
            "com.docker.network.bridge.host_binding_ipv4": "127.0.0.1",
            "com.docker.network.bridge.name": "docker0",
            "com.docker.network.driver.mtu": "1500"
        },
        "Labels": {}
    }

docker network inspect testnet      
[
    {
        "Name": "testnet",
        "Id": "de799b4d57b28fffbe5671d0381f246c3cf9c7aab7c45c6fbad57cf20647c312",
        "Created": "2020-04-17T00:52:35.0736808Z",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "172.19.0.0/16",
                    "Gateway": "172.19.0.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {},
        "Options": {},
        "Labels": {}
    }
]
```

The fix is easy, something similar to https://github.com/moby/moby/blob/646626d3be75c3e893d08041d37917d0a6ba9046/daemon/daemon_unix.go#L980
but any idea on what the design decisions were - was the `ip` option supposed to be generic/specific to the default bridge ?

**Comment by cpuguy83** [comment]
`ip` is one of those flags that existed before networks were a thing... but when networks were added any network level configuration on the daemon was lumped in with the mantra "this only applies to the default network".

Given port publishing is not really network dependent, I think lumping this one in was a mistake that nobody really noticed.

**Comment by arjenlentz** [comment]
Thanks for following up on this and figuring it out @cpuguy83 @arkodg - much appreciated!
Hopefully we'll see a fix for this soon, that'll be great.

**Comment by arkodg** [comment]
@cpuguy83 its network dependent because we are performing a DNAT translation from `ip:port` to the the network specific container IP e.g.
```
/usr/sbin/iptables, [--wait -t nat -C DOCKER -p tcp -d 127.0.0.1 --dport 80 -j DNAT --to-destination 172.18.0.2:80 ! -I docker0
```
And I agree with you, its safer for custom networks to not inherit all daemon settings

@arjenlentz you should be able to add the same functionality by adding this network option `com.docker.network.bridge.host_binding_ipv4` when creating your custom network
details in https://docs.docker.com/engine/reference/commandline/network_create/

**Comment by cpuguy83** [comment]
>  its network dependent because we are performing a DNAT translation from ip:port to the the network specific container IP e.g.

Good point.

**Comment by R70YNS** [comment]
Affecting both docker-compose and directly running the container. Tested running the daemon via `sudo /usr/bin/dockerd --init --ip=xxx.xxx.xxx.xxx` and calling an endpoint from within the container, which resulted in a different IP returned to that passed into the daemon.

**Comment by nicomro** [comment]
> you should be able to add the same functionality by adding this network option `com.docker.network.bridge.host_binding_ipv4` when creating your custom network

Thank you for the workaround @arkodg ! You saved my life.
